components/exercises/chapter-5-oscar/YourOwnLoop/loop.tsx

```diff
@@ -0,0 +1,22 @@
+/**
+ * Executes a loop until a test condition is false.
+ * @param value - The initial value for the loop.
+ * @param test - A function that tests the value for the loop condition.
+ * @param update - A function that updates the value for each iteration of the loop.
+ * @param body - A function that is executed for each iteration of the loop.
+ */
+export function loop(
+  value: number,
+  test: (value: number) => boolean,
+  update: (value: number) => number,
+  body: (value: number) => void
+): void {
+  const loopIteration = () => {
+    if (test(value)) {
+      body(value);
+      value = update(value);
+      setInterval(loopIteration, 1000);
+    }
+  };
+  loopIteration();
+}
```

Comment:
The problem with timing is since the function is recursive, we are setting several `setInterval`, **causing the function to be called more times than expected**. Solution: either use setInterval only once on the first call, or use `setTimeout` (which only runs once) on each recursive call.

Another problem, when the counter is reset, is that **the interval is not cleared** so it will jump between 10 and whatever number it was showing before. Solution: when resetting countdown, you should call `clearInterval` or `clearTimeout`. When doing this, remember you need to pass to clearInterval the interval id returned when doing setInterval, like this.
```js
const myInterval = setInterval(() => console.log('hello'), 1000)

clearInterval(myInterval)
```

More examples [here](https://www.geeksforgeeks.org/javascript-cleartimeout-clearinterval-method/)

Reply:
noted I will create a ticket to solve this